src/librbd/cache/pwl/rwl/PmemManager.cc

```diff
@@ -0,0 +1,195 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "PmemManager.h"
+
+#undef dout_subsys
+#define dout_subsys ceph_subsys_rbd_pwl
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::cache::pwl::rwl::PmemManager: " \
+                           << " " <<  __func__ << ": "
+
+namespace librbd::cache::pwl::rwl {
+
+static inline uint64_t size_to_bit(uint64_t size) {
+  return size % MIN_WRITE_ALLOC_SIZE ? size / MIN_WRITE_ALLOC_SIZE + 1 :
+         size / MIN_WRITE_ALLOC_SIZE;
+}
+
+PmemDev::PmemDev(const char * path, const uint64_t size, CephContext *cct)
+  : m_lock(ceph::make_mutex(pwl::unique_lock_name(
+      "librbd::cache::pwl::PmemManager::m_lock", this))),
+    m_cct(cct)
+{
+  m_head_addr = create_dev(path, size);
+  if (!m_head_addr) {
+    lderr(m_cct) << "failed to create pmem device." << dendl;
+    return;
+  }
+  m_max_bit = m_mapped_len / MIN_WRITE_ALLOC_SIZE - 1;  // start from 0
+  ldout(m_cct, 5) << "successfully created pmem Device. "
+                  << " m_is_pmem: " << m_is_pmem
+                  << " m_head_addr: " << (void *)m_head_addr
+                  << " m_mapped_len: " << m_mapped_len
+                  << " m_max_bit: " << m_max_bit
+                  << dendl;
+}
+
+PmemDev::PmemDev(const char * path, CephContext *cct)
+  : m_lock(ceph::make_mutex(pwl::unique_lock_name(
+      "librbd::cache::pwl::PmemManager::m_lock", this))),
+    m_cct(cct)
+{
+  m_head_addr = open_dev(path);
+  if (!m_head_addr) {
+    lderr(m_cct) << "failed to create pmem device." << dendl;
+    return;
+  }
+  m_max_bit = m_mapped_len / MIN_WRITE_ALLOC_SIZE - 1;
+  ldout(m_cct, 5) << "successfully map existing pmem Device. "
+                  << " m_is_pmem: " << m_is_pmem
+                  << " m_head_addr: " << (void *)m_head_addr
+                  << " m_mapped_len: " << m_mapped_len
+                  << " m_max_bit: " << m_max_bit
+                  << dendl;
+}
+
+PmemDev::~PmemDev() {
+}
+
+std::unique_ptr<PmemDev> PmemDev::pmem_create_dev(const char *path,
+    const uint64_t size, CephContext *cct) {
+  std::unique_ptr<PmemDev> pmem_dev;
+  pmem_dev =std::make_unique<PmemDev>(path, size, cct);
+  if (!pmem_dev.get()) {
```

Comment:
under what circumstances, `pmem_dev` does not contain a valid pointer? do you meant to use `std::nothrow`?

Reply:
> under what circumstances, `pmem_dev` does not contain a valid pointer? do you meant to use `std::nothrow`?

This is an error. I originally meant to judge the failure of the device creation, but whether the creation of the device was successful or not cannot be judged by the existence of this pointer. It should be judged by the head address of the created device. Please check the updated code, thanks.